File: zig/src/data_structures/singly_linked_list.zig
const std = @import("std");
const Allocator = std.mem.Allocator;

/// Singly linked list with forward traversal.
/// Time: O(1) for prepend, O(n) for other operations
/// Space: O(n)
pub fn SinglyLinkedList(comptime T: type) type {
    return struct {
        const Self = @This();
        const Node = struct {
            value: T,
            next: ?*Node,
        };

        allocator: Allocator,
        head: ?*Node,
        count: usize,

        pub fn init(allocator: Allocator) Self {
            return .{
                .allocator = allocator,
                .head = null,
                .count = 0,
            };
        }

        pub fn deinit(self: *Self) void {
            var current = self.head;
            while (current) |node| {
                const next = node.next;
                self.allocator.destroy(node);
                current = next;
            }
        }

        pub fn prepend(self: *Self, value: T) !void {
            const node = try self.allocator.create(Node);
            node.* = .{ .value = value, .next = self.head };
            self.head = node;
            self.count += 1;
        }

        pub fn append(self: *Self, value: T) !void {
            const node = try self.allocator.create(Node);
            node.* = .{ .value = value, .next = null };

            if (self.head == null) {
                self.head = node;
            } else {
                var current = self.head;
                while (current.?.next) |next| {
                    current = next;
                }
                current.?.next = node;
            }
            self.count += 1;
        }

        pub fn get(self: *const Self, index: usize) ?T {
            if (index >= self.count) return null;
            var current = self.head;
            var i: usize = 0;
            while (current) |node| : (i += 1) {
                if (i == index) return node.value;
                current = node.next;
            }
            return null;
        }

        pub fn removeAt(self: *Self, index: usize) ?T {
            if (index >= self.count) return null;

            if (index == 0) {
                const node = self.head.?;
                const value = node.value;
                self.head = node.next;
                self.allocator.destroy(node);
                self.count -= 1;
                return value;
            }

            var current = self.head;
            var i: usize = 0;
            while (current) |node| : (i += 1) {
                if (i == index - 1) {
                    const to_remove = node.next.?;
                    const value = to_remove.value;
                    node.next = to_remove.next;
                    self.allocator.destroy(to_remove);
                    self.count -= 1;
                    return value;
                }
                current = node.next;
            }
            return null;
        }

        pub fn indexOf(self: *const Self, value: T) ?usize {
            var current = self.head;
            var i: usize = 0;
            while (current) |node| : (i += 1) {
                if (node.value == value) return i;
                current = node.next;
            }
            return null;
        }

        pub fn contains(self: *const Self, value: T) bool {
            return self.indexOf(value) != null;
        }

        pub fn isEmpty(self: *const Self) bool {
            return self.count == 0;
        }

        pub fn size(self: *const Self) usize {
            return self.count;
        }
    };
}

test "SinglyLinkedList operations" {
    const allocator = std.testing.allocator;
    var list = SinglyLinkedList(i32).init(allocator);
    defer list.deinit();

    try list.prepend(2);
    try list.prepend(1);
    try list.append(3);

    try std.testing.expectEqual(@as(?i32, 1), list.get(0));
    try std.testing.expectEqual(@as(?i32, 2), list.get(1));
    try std.testing.expectEqual(@as(?i32, 3), list.get(2));
}

File: ruby/lib/data_structures/singly_linked_list.rb
# frozen_string_literal: true

module DSA
  module DataStructures
    # Singly linked list with forward traversal.
    # Time: O(1) for prepend, O(n) for other operations
    # Space: O(n)
    class SinglyLinkedList
      Node = Struct.new(:value, :next_node)

      def initialize
        @head = nil
        @size = 0
      end

      def prepend(value)
        @head = Node.new(value, @head)
        @size += 1
        self
      end

      def append(value)
        node = Node.new(value, nil)
        if @head.nil?
          @head = node
        else
          current = @head
          current = current.next_node while current.next_node
          current.next_node = node
        end
        @size += 1
        self
      end

      def insert(index, value)
        return false if index < 0 || index > @size

        if index.zero?
          prepend(value)
          return true
        end

        current = @head
        (index - 1).times { current = current.next_node }
        current.next_node = Node.new(value, current.next_node)
        @size += 1
        true
      end

      def remove_at(index)
        return nil if index < 0 || index >= @size

        if index.zero?
          value = @head.value
          @head = @head.next_node
          @size -= 1
          return value
        end

        current = @head
        (index - 1).times { current = current.next_node }
        value = current.next_node.value
        current.next_node = current.next_node.next_node
        @size -= 1
        value
      end

      def get(index)
        return nil if index < 0 || index >= @size

        current = @head
        index.times { current = current.next_node }
        current.value
      end

      def index_of(value)
        current = @head
        index = 0
        while current
          return index if current.value == value

          current = current.next_node
          index += 1
        end
        -1
      end

      def contains?(value)
        index_of(value) != -1
      end

      def empty?
        @size.zero?
      end

      def size
        @size
      end

      def clear
        @head = nil
        @size = 0
        self
      end

      def to_a
        result = []
        current = @head
        while current
          result << current.value
          current = current.next_node
        end
        result
      end
    end
  end
end
